### PR TITLE
User model tests

### PR DESCRIPTION
**Pull request by mayasa3**
User model tests:

- User creation
- Add user to database
- Check if User and UserAccount instances are created together(aka no users without user accounts or user accounts without users)
- Check if a UserAccount model instance is created when creating a User instance(testing the model inheritance/signals)

### CODE REVIEW

**Comment by mayasa3** [comment]
#noqa comments are to silence flake8. 
flake8 shows errors(line imported but not used), but tests won't work with the lines removed.

File: accounts/tests.py [hunk 1]
```diff
@@ -1,3 +1,43 @@
-# from django.test import TestCase
+import pytest
+from django.contrib.auth.models import User # noqa
+from .models import UserAccount
+from . import signals # noqa
```
Comment (benipeled) [inline accounts/tests.py]: Is `signals` also required? I can't see where it's used.

File: accounts/tests.py [hunk 2]
```diff
@@ -1,3 +1,43 @@
...
-# Create your tests here.
+
+@pytest.fixture
+def test_account(db):
+    username = 'testuser'
+    password = 'lvkmdgv32fdbf1b'
+    email = 'testuser@test.com'
+    new_user = User.objects.create_user(username=username,
+                                        password=password,
+                                        email=email)
```
Comment (benipeled) [inline accounts/tests.py]: You might want to use:
```
    new_user = User.objects.create_user(
        username=username, password=password, email=email
    )
```

File: accounts/tests.py [hunk 3]
```diff
@@ -1,3 +1,43 @@
...
+    return new_user
+
+
+@pytest.mark.django_db()
+def test_account_create(test_account):
+    test_account
```
Comment (benipeled) [inline accounts/tests.py]: What's the reason for this `test_account`? (same for all other tests)

**Review by benipeled — COMMENTED** [review]
In such cases, when you know the username/email of the added user, it's better to verify that the user is added by looking for the username/email of the added user, 

In addition, some of your tests are repetitive and actually test the same thing (existing user) - `User.objects.count()`.

**Comment by benipeled** [comment]
Two little tips:

- You can use `black` [1] to format your code - it has integration with pycharm and visual studio code (and other IDEs).
- You can organize the imports lines with `Ctrl+Alt+o` on pycharm, and something like [2] on visual studio code.

[1] https://pypi.org/project/black/
[2] https://stackoverflow.com/a/46722805

File: accounts/tests.py (hunk 1 above)
Comment (mayasa3) [reply accounts/tests.py]: the models file needs it, for some reason when I tried to run tests I had errors unless I imported it. Should I try removing the file and putting the signals in the models file?

File: accounts/tests.py (hunk 3 above)
Comment (mayasa3) [reply accounts/tests.py]: Originally it was supposed to be for less redundancy(creating a test user with the same details every time) but I'll remove it and try something else

File: accounts/tests.py (hunk 1 above)
Comment (benipeled) [reply accounts/tests.py]: I see,
@ifireball do you have an idea what's cause this required import?

The `signals` file contains the following that probably in use by the test somehow but I'm not sure how:
```
from django.contrib.auth.models import User
from django.db.models.signals import post_save
from django.dispatch import receiver
from .models import UserAccount


@receiver(post_save, sender=User)
def create_user_account(sender, instance, created, **kwargs):
    if created:
        UserAccount.objects.create(user=instance)


@receiver(post_save, sender=User)
def save_user_account(sender, instance, **kwargs):
    instance.useraccount.save()

```

File: accounts/tests.py (hunk 3 above)
Comment (benipeled) [reply accounts/tests.py]: It's already done by calling the fixture in the parentheses
`def test_account_create(test_account):`

File: accounts/tests.py [hunk 4]
```diff
@@ -1,3 +1,43 @@
-# from django.test import TestCase
+import pytest
+from django.contrib.auth.models import User # noqa
```
Comment (benipeled) [inline accounts/tests.py]: You can remove the `# noqa` from this line - it's in use later in the tests and I see no flake8 or other errors without it.

File: accounts/tests.py (hunk 1 above)
Comment (mayasa3) [reply accounts/tests.py]: One of the things I'm testing involves the signals, could it be that?
They're meant to create a connected UserAccount instance whenever a regular user is created, one thing I'm testing is whether that happens
EDIT: I tried moving the signals to models.py and removing signals.py, the tests pass with no flake8 issues but I'll make sure changing that doesn't mess anything else up

File: accounts/tests.py [hunk 5]
```diff
@@ -1,3 +1,33 @@
...
+def test_user_create(test_account):
+    assert User.objects.count() > 0
+    assert UserAccount.objects.count() > 0
+    assert UserAccount.objects.count() == User.objects.count()
+
+
+@pytest.mark.django_db()
+def test_user_email(test_account):
+    assert hasattr(test_account, 'email')
+    assert len(test_account.email) > 0
```
Comment (benipeled) [inline accounts/tests.py]: Both tests checking the test_account object but nothing with the created user on DB,
what about `assert User.objects.get(email=test_account.email)`?

Same for `test_user_find` you can use the `User.objects.get` to find the created user.

The `<modelName>.objects.get(<some_filter>)` [1] can help with compare/verify db-content 
[1] https://docs.djangoproject.com/en/3.1/topics/db/queries/#retrieving-specific-objects-with-filters

**Comment by benipeled** [comment]
In addition, please add a short docstring for each test - no need something detailed, short explanation is enough.

File: accounts/tests.py [hunk 6]
```diff
@@ -1,3 +1,33 @@
...
+    email = 'testuser@test.com'
+    new_user = User.objects.create_user(
+        username=username, password=password, email=email
+    )
+    return new_user
+
+
+@pytest.mark.django_db()
+def test_user_create(test_account):
+    assert User.objects.count() > 0
```
Comment (bardielle) [inline accounts/tests.py]: we discussed about that- it not a good validation 
please asserts the values

File: accounts/tests.py (hunk 5 above)
Comment (bardielle) [reply accounts/tests.py]: +1

**Comment by mayasa3** [comment]
Changed the tests, does it look better?

File: accounts/tests.py [hunk 7]
```diff
@@ -1,3 +1,30 @@
...
+    username = 'testuser'
+    password = 'lvkmdgv32fdbf1b'
+    email = 'testuser@test.com'
+    new_user = User.objects.create_user(
+        username=username, password=password, email=email
+    )
+    return new_user
+
+
+# Create a user and check if a UserAccount instance was created
```
Comment (benipeled) [inline accounts/tests.py]: This docstring (comment) should be within the function
```
def some_function():
    """This is an example function with docstring"""
```

You can read more about docstring on https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

File: accounts/tests.py [hunk 8]
```diff
@@ -1,3 +1,30 @@
...
+@pytest.mark.django_db()
+def test_user_create(test_account):
+    assert User.objects.count() == 1
+    assert UserAccount.objects.count() == 1
+
+
+# Find user in database by user id, username or email
+@pytest.mark.django_db()
+def test_user_find(test_account):
+    assert User.objects.get(id=test_account.id)
```
Comment (benipeled) [inline accounts/tests.py]: You can skip this test, the next two (username and email) should be enough

Comment (mayasa3) [reply accounts/tests.py]: Removed it and moved the comments

**Comment by benipeled** [comment]
LGTM,
@bardielle if you have nothing else to add - it can be merged.

**Comment by guy9050** [comment]
lgtm

File: accounts/tests.py [hunk 9]
```diff
@@ -1,3 +1,29 @@
-# from django.test import TestCase
+import pytest
+from django.contrib.auth.models import User
+from .models import UserAccount
 
-# Create your tests here.
+
+@pytest.fixture
+def test_account(db):
```
Comment (guy9050) [inline accounts/tests.py]: Fixtures names should not start with "test_", only actual tests start like that.
You can call this fixture "user_account" or something like that.

File: accounts/tests.py [hunk 10]
```diff
@@ -1,3 +1,29 @@
...
+    new_user = User.objects.create_user(
+        username=username, password=password, email=email
+    )
+    return new_user
+
+
+@pytest.mark.django_db()
+def test_user_create(test_account):
+    """Create a user and check if a UserAccount instance was created"""
+    assert User.objects.count() == 1
```
Comment (guy9050) [inline accounts/tests.py]: @bardielle is it okay to compare the number of results? I think you said not to do it

File: accounts/tests.py [hunk 11]
```diff
@@ -1,3 +1,29 @@
...
+    assert UserAccount.objects.count() == 1
+
+
+@pytest.mark.django_db()
+def test_user_find(test_account):
+    """Find user in database by user id, username or email"""
+    assert User.objects.get(username=test_account.username)
+    assert User.objects.get(email=test_account.email)
```
Comment (guy9050) [inline accounts/tests.py]: if you found that user by username there is no reason you wouldn't find it by email, so the second assert seems redundant. @benipeled @bardielle thoughts?

File: accounts/tests.py (hunk 10 above)
Comment (mayasa3) [reply accounts/tests.py]: Changed it to UserAccount.objects.get(user=user_account). If it wasn't created, the test wouldn't pass anyway, so I think it works? Correct me if I'm wrong though.

File: accounts/tests.py [hunk 12]
```diff
@@ -1,3 +1,28 @@
-# from django.test import TestCase
+import pytest
+from django.contrib.auth.models import User
+from .models import UserAccount
 
-# Create your tests here.
+
+@pytest.fixture
```
Comment (guy9050) [inline accounts/tests.py:6]: @ifireball is it okay to create fixtures here, or should be under conftest.py?

Comment (ifireball) [reply accounts/tests.py:6]: `conftest.py` is for sharing fixtures between multiple modules (files), if the fixtures aren't shared they should not be there.

Comment (bardielle) [reply accounts/tests.py:6]: @mayasa3 is it shared fixture? it seems that it isn't

File: accounts/tests.py (hunk 10 above)
Comment (bardielle) [reply accounts/tests.py]: so you don't need that line

File: accounts/tests.py (hunk 12 above)
Comment (mayasa3) [reply accounts/tests.py:6]: It isn't shared, I just added it for my tests for the user model. I don't know if it's relevant for the others, but they haven't used it.

File: accounts/tests.py (hunk 10 above)
Comment (mayasa3) [reply accounts/tests.py]: The line that compares them or what I changed it to?  Should I remove that too?

Comment (guy9050) [reply accounts/tests.py]: you added line 22, so line 21 is not needed

Comment (mayasa3) [reply accounts/tests.py:21]: Then what way would be better if I want to test the model inheritance? Or do I not need a test for that at all?

**Review by bardielle — APPROVED** [review]
/lgtm

**Pull request merged**